posts/cocurrency-parallelism-in-elixir.md

```diff
@@ -0,0 +1,173 @@
+Exercises on Exercism are small, synthetic, and often seemingly trivial. It’s easy to imagine that experienced practitioners would have nothing to learn from them. However, solving these synthetic problems can push you to learn and apply parts of your language that you may not have explored. This new learning can lead you to solve real world problems more efficiently or in a more expressive way.
+
+[Parallel Letter Frequency](https://exercism.io/tracks/elixir/exercises/parallel-letter-frequency) is a medium difficulty exercise on [Exercism's Elixir Track](https://exercism.io/tracks/elixir) that unpacks a surprising number of interesting lessons. To solve this problem successfully, your solution should execute in parallel in a number of worker processes. Achieving true parallelism in Elixir is surprisingly easy compared to other languages, but if you've never written concurrent code in Elixir, it may seem a little daunting. One of the things you'll discover in solving this exercise is how easy Elixir makes it to write code that can execute concurrently or in parallel. Applying these skills to your own code can have a big impact on the performance of your applications.
+
+This exercise requires you to implement a `Frequency.frequency/2` function that determines the letter frequency in a list of strings. The calculation should be carried out in a number of worker processes, set by the `workers` argument:
+
+```elixir
+iex> Frequency.frequency(["Freude", "schöner", "Götterfunken"], workers)
+%{
+  "c" => 1, 
+  "d" => 1, 
+  "e" => 5, 
+  ...
+  "ö" => 2
+}
+```
+
+In this post we'll explore concurrency in Elixir by making a working sequential solution to this exercise concurrent. But before we jump into the code, let's take a moment to examine what "concurrency" and "parallelism" actually mean, and how to achieve both in Elixir vs other languages.
+
+## Concurrency and parallelism
+
+Concurrency and parallelism are related terms, but don't mean exactly the same thing. Both terms describe how tasks are executed by a program:
+
+* A _concurrent_ program is one where multiple tasks can be "in progress", but at any single point in time, only one task is executing on the CPU (e.g. executing one task while another is waiting for IO [disk, network, etc.]).
+* A _parallel_ program on the other hand is capable of executing multiple tasks _at the same time_ on multiple CPU cores.
+
+Both concurrent and parallel execution can give major speed increases, but the situations where this is true (or even possible) are dependent on a multitude of factors, some of which are: 
```

Comment:
I'm not certain about this one, but it might be easier to digest this bit if we move away from the bullet point list. E.g.:

> Both concurrent and parallel execution can give major speed increases, but the situations where this is true (or even possible) are dependent on a multitude of factors, such how many CPUs are available, and whether or not it's even possible to do the tasks concurrently or in parallel.

Then maybe follow on with something that is a bit more discussion based, to explore a couple of the follow-up points that you have. E.g.:

> If you've got just one CPU core, then you're almost certainly not going to be able to get parallel execution, but even then, if your tasks are IO-bound, then you might get speedup even on a single core by making the tasks concurrent. [It's not obvious why this is true, so it could be worth talking about why that is, then]. If you then add parallelism to these concurrent, IO-bound tasks then you probably won't see much additional improvement.

Then maybe a similar paragraph/discussion about things that are CPU-bound.

Reply:
Good points, I'll try and improve this area.